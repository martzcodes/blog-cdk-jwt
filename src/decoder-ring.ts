import {
  APIGatewayProxyEvent,
  APIGatewayProxyResult,
// eslint-disable-next-line import/no-unresolved
} from 'aws-lambda';
import { decode, verify } from 'jsonwebtoken';
import jwksClient, { JwksClient } from 'jwks-rsa';


interface Body {
  jwksUri: string;
  token: string;
}

export const getSigningKey = async (client: JwksClient, kid: string): Promise<jwksClient.SigningKey> => {
  return new Promise<jwksClient.SigningKey>((res, rej) => {
    client.getSigningKey(kid, (err, key) => {
      if (err) {
        rej(err);
      } else {
        res(key);
      }
    });
  });
};

export const verifyToken = async (client: JwksClient, token: string) => {
  const decoded = decode(token, { complete: true, json: true });
  if (!decoded || !decoded.header.kid) {
    console.error('No Key ID');
    throw new Error('Token has no Key ID');
  }

  const key = await getSigningKey(client, decoded.header.kid);
  const signingKey = key.getPublicKey();

  if (!signingKey) {
    console.error('No public or rsa key');
    throw new Error('Unauthorized');
  }

  return verify(token, signingKey);
};

export const handler = async (
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
  if (!event.body) {
    return {
      statusCode: 400,
      body: 'Bad Payload',
    };
  }

  const body: Body = JSON.parse(event.body) as Body;

  const client = jwksClient({
    cache: true,
    rateLimit: true,
    jwksRequestsPerMinute: 10,
    jwksUri: body.jwksUri,
  });

  try {
    const verified = await verifyToken(client, body.token);


    return {
      statusCode: 200,
      body: JSON.stringify(verified),
    };
  } catch (e) {
    console.error(e);
    return {
      statusCode: 500,
      body: e.message,
    };
  }
};