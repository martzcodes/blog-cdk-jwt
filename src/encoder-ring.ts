// eslint-disable-next-line import/no-unresolved
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import SecretsManager from 'aws-sdk/clients/secretsmanager';
import { JWK, JWS } from 'node-jose';

const sm = new SecretsManager({ region: `${process.env.AWS_REGION}` });

export const signToken = async (tokenPayload: any, keyBuffer: Buffer) => {
  const key = await JWK.asKey(keyBuffer);

  const signOptions = { compact: true, fields: { typ: 'jwt' } };
  const currentTime = Math.floor(Date.now() / 1000);
  const payload = {
    exp: currentTime + 86400, // 1 Day from now
    iat: currentTime,
    sub: tokenPayload.userId,
    ...tokenPayload,
  };
  return (await JWS.createSign(signOptions, key).update(JSON.stringify(payload)).final()).toString();
};

export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
  if (!event.body) {
    return {
      statusCode: 400,
      body: 'Bad Payload',
    };
  }

  try {
    const tokenPayload = JSON.parse(event.body);

    const secret = await sm.getSecretValue({ SecretId: `${process.env.SECRET_ID}` }).promise();
    const jwks = JSON.parse(`${secret.SecretString}`);
    const token = await signToken(tokenPayload, jwks);

    return {
      statusCode: 200,
      body: JSON.stringify({ token }),
    };
  } catch (error) {
    console.error(error.message);

    return {
      statusCode: 500,
      body: 'InternalServerError',
    };
  }
};