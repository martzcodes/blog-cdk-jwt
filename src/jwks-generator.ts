import S3 from 'aws-sdk/clients/s3';
import SecretsManager from 'aws-sdk/clients/secretsmanager';
import { JWK } from 'node-jose';

const sm = new SecretsManager({ region: `${process.env.AWS_REGION}` });
const s3 = new S3();

export const handler = async (): Promise<void> => {
  // when invoked this regenerates the jwks and stores the secret in secrets manager and public in s3
  try {
    const keyStore = JWK.createKeyStore();
    const key = await keyStore.generate('RSA', 2048, {
      alg: 'RS256',
      use: 'sig',
    });

    const secret = key.toJSON(true);
    await sm
      .putSecretValue({
        SecretId: `${process.env.SECRET_ID}`,
        SecretString: JSON.stringify(secret),
      })
      .promise();

    const publicKey = key.toJSON();
    await s3
      .putObject({
        Bucket: `${process.env.BUCKET_ID}`,
        Key: 'jwks.json',
        Body: JSON.stringify(publicKey),
        ACL: 'bucket-owner-full-control',
      })
      .promise();
  } catch (e) {
    console.error(e);
    throw e;
  }
};