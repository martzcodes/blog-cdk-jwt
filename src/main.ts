import { join } from 'path';
import { AwsIntegration, IntegrationResponse, LambdaIntegration, RestApi } from '@aws-cdk/aws-apigateway';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Runtime } from '@aws-cdk/aws-lambda';
import { LogLevel, NodejsFunction, NodejsFunctionProps } from '@aws-cdk/aws-lambda-nodejs';
import { BlockPublicAccess, Bucket, BucketAccessControl } from '@aws-cdk/aws-s3';
import { Secret } from '@aws-cdk/aws-secretsmanager';
import { App, Construct, CustomResource, Duration, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';
import { Provider } from '@aws-cdk/custom-resources';

export class BlogCdkJwksStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const SECRET_ID = 'BlogCdkSecret';
    const secret = new Secret(this, SECRET_ID);

    const BUCKET_ID = 'BlogCdkJwksBucket';
    const jwksBucket = new Bucket(this, BUCKET_ID, {
      accessControl: BucketAccessControl.BUCKET_OWNER_FULL_CONTROL,
      bucketName: `blog-cdk-jwks-bucket-${this.account}`,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });
    const bucketRole = new Role(this, 'BlogCdkJwksBucketRole', {
      roleName: 'BlogCdkJwksBucketRole',
      assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
    });
    jwksBucket.grantRead(bucketRole);

    const lambdaProps: Partial<NodejsFunctionProps> = {
      bundling: {
        target: 'es2020',
        logLevel: LogLevel.ERROR,
      },
      runtime: Runtime.NODEJS_14_X,
      environment: {
        NODE_ENV: process.env.NODE_ENV as string,
        SECRET_ID: secret.secretArn,
        BUCKET_ID: jwksBucket.bucketName,
      },
    };

    const jwksGenerator = new NodejsFunction(this, 'BlogCdkJwksGenerator', {
      ...lambdaProps,
      timeout: Duration.seconds(30),
      entry: join(__dirname, './jwks-generator.ts'),
    });
    secret.grantWrite(jwksGenerator);
    jwksBucket.grantWrite(jwksGenerator);

    const jwksGeneratorProvider = new Provider(this, 'BlogCdkJwksGenerateProvider', {
      onEventHandler: jwksGenerator,
    });

    new CustomResource(this, 'BlogCdkJwksGenerateResource', {
      serviceToken: jwksGeneratorProvider.serviceToken,
      properties: {
        // Bump to force an update
        Version: '2',
      },
    });

    const restApi = new RestApi(this, 'BlogCdkJwksApi');

    const jwksIntegration = new AwsIntegration({
      service: 's3',
      integrationHttpMethod: 'GET',
      path: `${jwksBucket.bucketName}/jwks.json`,
      options: {
        credentialsRole: bucketRole,
        // integration responses are required!
        integrationResponses: [
          {
            'statusCode': '200',
            'method.response.header.Content-Type': 'integration.response.header.Content-Type',
            'method.response.header.Content-Disposition': 'integration.response.header.Content-Disposition',
          } as IntegrationResponse,
          { statusCode: '400' },
        ],
      },
    });

    restApi.root
      .addResource('.well-known')
      .addResource('jwks.json')
      .addMethod('GET', jwksIntegration, {
        methodResponses: [{ statusCode: '200' }],
      });

    const encoderFunction = new NodejsFunction(this, 'BlogCdkJwtEncoderFn', {
      ...lambdaProps,
      entry: join(__dirname, './encoder-ring.ts'),
    });
    secret.grantRead(encoderFunction);

    restApi.root.addResource('encode').addMethod('POST', new LambdaIntegration(encoderFunction));

    const decoderFunction = new NodejsFunction(this, 'BlogCdkJwtDecoderFn', {
      ...lambdaProps,
      entry: join(__dirname, './decoder-ring.ts'),
    });

    restApi.root.addResource('decode').addMethod('POST', new LambdaIntegration(decoderFunction));
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new BlogCdkJwksStack(app, 'blog-cdk-jwks-dev', { env: devEnv });

app.synth();