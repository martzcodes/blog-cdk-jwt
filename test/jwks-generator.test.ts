import S3 from '../src/__mocks__/aws-sdk/clients/s3';
import SecretsManager, { mockSMResponse } from '../src/__mocks__/aws-sdk/clients/secretsmanager';
import { handler } from '../src/jwks-generator';

const s3 = new S3();
const sm = new SecretsManager();

describe('JwksGenerator', () => {
  test('generate jwks', async () => {
    await handler();
    expect(sm.putSecretValue).toHaveBeenCalledTimes(1);
    expect(s3.putObject).toHaveBeenCalledTimes(1);
    const putSecretValue = sm.putSecretValue.mock.calls[0][0];
    expect(putSecretValue).toMatchSnapshot({
      SecretString: expect.any(String),
    });
    const secretValue = JSON.parse(putSecretValue.SecretString);
    expect(secretValue).toMatchSnapshot({
      kty: 'RSA',
      kid: expect.any(String),
      use: 'sig',
      alg: 'RS256',
      e: expect.any(String),
      n: expect.any(String),
      d: expect.any(String),
      p: expect.any(String),
      q: expect.any(String),
      dp: expect.any(String),
      dq: expect.any(String),
      qi: expect.any(String),
    });

    const putObject = s3.putObject.mock.calls[0][0];
    expect(putObject).toMatchSnapshot({
      Body: expect.any(String),
    });
    const body = JSON.parse(putObject.Body);
    expect(body).toMatchSnapshot({
      kty: 'RSA',
      kid: expect.any(String),
      use: 'sig',
      alg: 'RS256',
      e: expect.any(String),
      n: expect.any(String),
    });
  });
  test('generate jwks error', async () => {
    const errorMessage = 'some error';
    mockSMResponse.mockRejectedValue(new Error(errorMessage));
    expect.assertions(1);
    try {
      await handler();
    } catch (e) {
      console.log(e);
      expect(e.message).toBe(errorMessage);
    }
  });
});