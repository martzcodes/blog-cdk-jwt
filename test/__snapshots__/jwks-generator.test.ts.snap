// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JwksGenerator generate jwks 1`] = `
Object {
  "SecretId": "undefined",
  "SecretString": Any<String>,
}
`;

exports[`JwksGenerator generate jwks 2`] = `
Object {
  "alg": "RS256",
  "d": Any<String>,
  "dp": Any<String>,
  "dq": Any<String>,
  "e": Any<String>,
  "kid": Any<String>,
  "kty": "RSA",
  "n": Any<String>,
  "p": Any<String>,
  "q": Any<String>,
  "qi": Any<String>,
  "use": "sig",
}
`;

exports[`JwksGenerator generate jwks 3`] = `
Object {
  "ACL": "bucket-owner-full-control",
  "Body": Any<String>,
  "Bucket": "undefined",
  "Key": "jwks.json",
}
`;

exports[`JwksGenerator generate jwks 4`] = `
Object {
  "alg": "RS256",
  "e": Any<String>,
  "kid": Any<String>,
  "kty": "RSA",
  "n": Any<String>,
  "use": "sig",
}
`;
